driver-dse/src/main/java/com/datastax/driver/auth/DseAuthProvider.java

```diff
@@ -65,20 +91,153 @@
+        // Need both username and password for PLAIN auth
+        if ((username != null) && (password != null)) {
+            return new PlainTextAuthenticator(authenticator, username, password);
+        }
+        return new GSSAPIAuthenticator(authenticator, host);
+    }
+
+    public static abstract class BaseDseAuthenticator implements Authenticator {
+        private final String authenticator;
+
+        public BaseDseAuthenticator(String authenticator) {
+            this.authenticator = authenticator;
+        }
+
+        public abstract byte[] getMechanism();
+        public abstract byte[] getInitialServerChallenge();
+
+        public byte[] initialResponse() {
+            if (isDseAuthenticator())
+                return getMechanism();
+            else
+                return evaluateChallenge(getInitialServerChallenge());
+        }
+
+        public void onAuthenticationSuccess(byte[] token) {
+        }
+
+        private boolean isDseAuthenticator() {
+            return authenticator.equals(DSE_AUTHENTICATOR);
+        }
+    }
+
+    private static class PlainTextAuthenticator extends BaseDseAuthenticator {
+        private static final byte[] MECHANISM = "PLAIN".getBytes(Charsets.UTF_8);
+        private static final byte[] SERVER_INITIAL_CHALLENGE = "PLAIN-START".getBytes(Charsets.UTF_8);
+        private final byte[] username;
+        private final byte[] password;
+
+        public PlainTextAuthenticator(String authenticator, String username, String password) {
+            super(authenticator);
+            this.username = username.getBytes(Charsets.UTF_8);
+            this.password = password.getBytes(Charsets.UTF_8);
+        }
+
+        public byte[] getMechanism() {
+            return MECHANISM.clone();
+        }
+
+        public byte[] getInitialServerChallenge() {
+            return SERVER_INITIAL_CHALLENGE.clone();
+        }
+
+        public byte[] evaluateChallenge(byte[] challenge) {
+            if (Arrays.equals(SERVER_INITIAL_CHALLENGE, challenge)) {
+                byte[] token = new byte[username.length + password.length + 2];
+                token[0] = 0;
+                System.arraycopy(username, 0, token, 1, username.length);
+                token[username.length + 1] = 0;
+                System.arraycopy(password, 0, token, username.length + 2, password.length);
+                return token;
+            }
+            throw new RuntimeException("Incorrect challenge from server");
+        }
+    }
+
+    private static class GSSAPIAuthenticator extends BaseDseAuthenticator {
+        private static final String JAAS_CONFIG_ENTRY = "DseClient";
+        private static final String[] SUPPORTED_MECHANISMS = new String[]{"GSSAPI"};
+        private static final String SASL_PROTOCOL_NAME = "dse";
+        private static final String SASL_PROTOCOL_NAME_PROPERTY = "dse.sasl.protocol";
+        private static final Map<String, String> DEFAULT_PROPERTIES = new HashMap<String, String>()
```

Comment:
This does not look thread-safe, maybe replace with `ImmutableMap.<String, String>builder().put(Sasl.SERVER_AUTH, "true").put(Sasl.QOP, "auth").build()`?